streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -0,0 +1,306 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+class NamedCache {
+    private static final Logger log = LoggerFactory.getLogger(NamedCache.class);
+    private final String name;
+    private final TreeMap<Bytes, LRUNode> cache = new TreeMap<>();
+    private final Set<Bytes> dirtyKeys = new LinkedHashSet<>();
+    private ThreadCache.DirtyEntryFlushListener listener;
+    private LRUNode tail;
+    private LRUNode head;
+    private long currentSizeBytes;
+    private ThreadCacheMetrics metrics;
+
+    // JMX stats
+    private Sensor hits = null;
```

Comment:
Instead of reporting hit miss rates (that is my assumption, see comments above), and overwrite / flush rates, I'd suggest we expose just a single metric first: 1) the hit ratio (not rate). 

Again, adding more metrics later are easier than removing or changing semantics on some of the existing ones.